presto-main/src/main/java/com/facebook/presto/cost/ComposableStatsCalculator.java

```diff
@@ -66,21 +66,26 @@ public ComposableStatsCalculator(List<Rule> rules)
     @Override
     public PlanNodeStatsEstimate calculateStats(PlanNode node, StatsProvider sourceStats, Lookup lookup, Session session, Map<Symbol, Type> types)
     {
-        Iterator<Rule> ruleIterator = getCandidates(node).iterator();
+        Iterator<Rule<?>> ruleIterator = getCandidates(node).iterator();
         while (ruleIterator.hasNext()) {
-            Rule rule = ruleIterator.next();
-            Optional<PlanNodeStatsEstimate> calculatedStats = rule.calculate(node, sourceStats, lookup, session, types);
+            Rule<?> rule = ruleIterator.next();
+            Optional<PlanNodeStatsEstimate> calculatedStats = calculatedStats(rule, node, sourceStats, lookup, session, types);
             if (calculatedStats.isPresent()) {
                 return calculatedStats.get();
             }
         }
         return PlanNodeStatsEstimate.UNKNOWN_STATS;
     }
 
-    public interface Rule
+    private static <T extends PlanNode> Optional<PlanNodeStatsEstimate> calculatedStats(Rule<T> rule, PlanNode node, StatsProvider sourceStats, Lookup lookup, Session session, Map<Symbol, Type> types)
     {
-        Pattern<? extends PlanNode> getPattern();
+        return rule.calculate((T) node, sourceStats, lookup, session, types);
```

Comment:
IDE shows a warning on this line. Since it's intentional, why don't you suppress it?

Reply:
@findepi I don't see any warnings. Do you also see warnings for `(T) object` in `com.facebook.presto.matching.DefaultMatcher`?

Reply:
This is intellij's `Java | Compiler issues | Unchecked warning` (i think it's enabled by default, but I might be wrong)

> Do you also see warnings for (T) object in com.facebook.presto.matching.DefaultMatcher?

yes

Reply:
@findepi I'm using IntelliJ 2017.3.4 (Community Edition). I don't have `Java | Compiler issues | Unchecked warning`. The closest view I found is this:

<img width="1023" alt="screen shot 2018-06-04 at 4 07 18 pm" src="https://user-images.githubusercontent.com/27965151/40938977-63a75b5e-6811-11e8-987b-a969ca4721d2.png">